Inference : Attempt to use CUDA before CPU

ericmehl opened the pull request:
This adds GPU support to the GaffML Inference node.

This isn't ready for review yet, I'm just submitting it to generate build artifacts for testing. There's probably some discussion to be had around how to expose this to users. For now it's hard coded to try CUDA and ONNX seems to fall back to the CPU if all the libraries it needs are not found.

There are a few things needed to get the GPU working :
1. Use the ONNX runtime release for your platform with `gpu` in the name. That will have `lib/onnxruntime_providers_cuda.dll` in the archive which is what onnxruntime needs.
2. Download and extract cuDNN : https://developer.nvidia.com/cudnn
3. Add the path to the extracted cuDNN `bin` directory to your `PATH` environment variable.
4. Install CUDA toolkit if you don't have it already : https://developer.nvidia.com/cuda-downloads
5. Linux might need Zlib? I've tested on Windows so not sure about that.

More instructions are at https://onnxruntime.ai/docs/install/

I found the following versions to work :
- ONNX runtime : 1.19.2
- cuDNN : 9.8.0
- CUDA toolkit : 12.8

### Checklist ###

- [ ] I have read the [contribution guidelines](https://github.com/GafferHQ/gaffer/blob/main/CONTRIBUTING.md).
- [ ] I have updated the documentation, if applicable.
- [ ] I have tested my change(s) in the test suite, and added new test cases where necessary.
- [ ] My code follows the Gaffer project's prevailing coding style and conventions.

johnhaddon commented:
> There's probably some discussion to be had around how to expose this to users. 

One thought I had was to do this with a context variable, like we do for the OCIO configuration. Then you can have a global setting on the script and override it locally if you need to. Not sure if that's overkill or not?

BrianHanke commented:
Hey Eric, I gave this a quick spin and when trying to load a model in Gaffer I get the error below. I have the versions of everything that you recommended and all environment variables look correct. Any idea what the problem might be?

ERROR   [Error loading model] RuntimeError: D:\a\_work\1\s\onnxruntime\core\session\provider_bridge_ort.cc:1637 onnxruntime::ProviderLibrary::Get [ONNXRuntimeError] : 1 : FAIL : LoadLibrary failed with error 126 "" when trying to load "C:\Source\lib\onnx\onnxruntime-win-x64-gpu-1.19.2\lib\onnxruntime_providers_cuda.dll"

boberfly commented:
I haven't jumped into this too much, but I was curious if anyone knows the minimal amount of CUDA toolkit you'd need as it's quite substantially sized for packaging up eg. for REZ, and things like Cycles tend to do a dynamic runtime link to the driver directly without the need for the toolkit for deployment...

ericmehl commented:
Hey @BrianHanke, great to have you trying this out! Getting all the environment setup and the right packages was the hardest part so it's great to have someone else testing it.

Just to double check, the error message is correctly pointing to your ONNX runtime at `C:\Source\lib\onnx\onnxruntime-win-x64-gpu-1.19.2\lib\onnxruntime_providers_cuda.dll`? If so, then at least it's finding that library, so it means the problem is one of those it depends on.

From within Gaffer, my `PATH` variable has entries for (with all of the $ variables expanded to their full path):
1. `$ONNX_ROOT/lib`
2. `$CUDNN_ROOT/bin`
3. `C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.8\bin` which was set up by the CUDA installer default
4. `C:\Program Files\NVIDIA GPU Computing Toolkit\CUDA\v12.8\libnvvp` from the CUDA installer

I also have an environment variable called `CUDA_PATH` that the installer set up. That may be needed too?

If those all look good, it may be time to get out the dependency walker. You can download that from here if you don't have it : https://www.dependencywalker.com/

Use `gaffer env depends.exe` to run the tool with the Gaffer environment set up. In that tool, load the `onnxruntime_providers_cuda.dll" file. It will probably take a minute or two to load, it's super slow. It will then show a tree of all the DLL files it's trying to load, and flag them as red or yellow if there are problems finding the file. For me, some of the core Windows libraries are flagged as a yellow warning, but those can usually be ignored.

Do you get a red error flag on any of the CUDA related libraries or sub-libraries in the tree?

BrianHanke commented:
@ericmehl Thanks, Dependency Walker took forever but it did identify the problem. I had C:\Source\lib\CUDNN\v9.8\bin in PATH and it turns out the CUDNN installer put all the files inside of bin\12.8. So, I added "12.8" to the end and all is working now! 

The bad news is that GPU is twice as slow as CPU in this 4x upscaler I have set up in Gaffer. It's about 40 seconds per tile instead of 20 seconds. I've got an RTX 2070. Not sure if that's an expected result. I assumed machine learning models work better on GPU, but I'm still learning about all this stuff.

ericmehl commented:
I reworked the commit to hopefully be a little bit easier and more resilient. Now, setting `GAFFERML_USE_CUDA=1` will cause GafferML to attempt to use CUDA. If it doesn't work, it will give an error and fallback to the CPU. Without that, it just uses the CPU.

I changed how to setup the cuDNN environment. Now you need to set `CUDNN_ROOT` to the directory you extracted cuDNN to. We'll add the `/bin` directory in the setup script. And it warns if you don't have that, so hopefully it's a small step toward making this more intuitive to setup.

I still haven't come up with a satisfactory way of telling users they need to also have the CUDA toolkit installed.

> The bad news is that GPU is twice as slow as CPU in this 4x upscaler I have set up in Gaffer

Well that's a big disappointment! I'm not experienced with all this stuff either so I don't have great troubleshooting advice for you. Does the Windows Task Manager show GPU activity while you're running it? I wonder if something isn't getting configured right. For me it takes about a second per 64 pixel tile on the GPU (for the 4x upscaler you linked to on Discord).

It sounds like Nvidia's _other_ execution provider, TensorRT, might give better performance. But it also requires passing the model through a converter to use it, which is annoying.

BrianHanke commented:
@ericmehl It's definitely using the GPU so that part is good! 

Did you try the latest upscaler I added to GitHub? upscalev_3.gfr and tile_v3.ps1 here: https://github.com/BrianHanke/GafferML plus a link to the model in the README. 

It's much more intensive than the first one I tried and gives much better results. A 1920x1080 image cut into into 121 192x108 tiles takes about an hour on CPU or two on GPU. I honestly don't know if GPU should be expected to perform better out of the box or if there are optimizations that need to be made.

BrianHanke commented:
Hey @ericmehl, whoa, major development here! 

I was testing things a bit more and totally by accident I noticed that an upscaled tile was loading INSTANTLY in the viewer. That's weird, I thought, so I went to run the Dispatcher to save all tiles to disk again. Nope, 40 seconds each. 

Instant in the Viewer, slow saving to disk... hmm... 

Then I had the thought that the Dispatcher must be like reloading the entire model every tile or something, I unchecked Execute in Background in the Dispatcher and it spit out all 121 tiles in TWELVE SECONDS!

So yeah, GPU is working spectacularly well for this task now. Was I just being super dumb with my original Dispatcher setup? I'm honestly not sure what's going on here and why what I did worked.

johnhaddon commented:
When the LocalDispatcher is executing in the background, it launches a new process for each task. So if each tile is a task, then each tile will indeed launch Gaffer and then load the ONNX model.

Not executing in the background is a decent short-term fix, but it's not what you'd want if you had a GPU farm. This is where nested dispatchers come in. You can place a not-executing-in-the-background LocalDispatcher below your tile tasks nodes, so that no matter what the outer dispatcher is (Deadline, Tractor etc), all those tasks will get done in a single Gaffer process. This works for local dispatch too - if your outer dispatcher is a LocalDispatcher in background mode, then you'll be launching a single background process to do all the tiles.

BrianHanke commented:
Thanks for the info John! Exciting to find this huge performance boost. An extremely high quality upscale in 10 seconds or so is a great deal. I checked to see if tiles are still necessary, but they are. VRAM runs out fast.

murraystevenson commented:
This is now rebased and retargeted to `main`, with a Changes.md entry added for 1.6, and the `CUDNN_ROOT` behaviour adjusted to also work on Linux. @ericmehl it'd be worth giving this another try on Windows to make sure I haven't bodged anything.

From my testing on Linux, this change is isolated enough that users who haven't set `GAFFERML_USE_CUDA` and are using the non-gpu ONNX runtime for inference shouldn't be adversely affected. 

As far as testing on Linux goes, I was able to run @BrianHanke's upscale setup on the GPU, with a 10x speed improvement over CPU inference when upscaling `macaw.exr`. I did see a number of onnxruntime warnings in the process, with a `onnxruntime::OnnxRuntimeException` thrown with a CUDA failure on Gaffer shutdown. Not sure if that is related to the driver/CUDA/cuDNN combination I'm running on my workstation, but I think it's fair to label this as experimental...

BrianHanke commented:
Same, I get a ton of ONNX warnings in the console when using GPU on Windows.
